crates/wasmparser/src/readers/core/types.rs

```diff
@@ -55,10 +55,56 @@ pub enum Type {
 /// Represents a type of a function in a WebAssembly module.
 #[derive(Debug, Clone, Eq, PartialEq, Hash)]
 pub struct FuncType {
-    /// The function parameter types.
-    pub params: Box<[ValType]>,
-    /// The function result types.
-    pub returns: Box<[ValType]>,
+    /// The combined parameters and result types.
+    params_results: Box<[ValType]>,
+    /// The number of paramter types.
+    len_params: usize,
+}
+
+impl FuncType {
+    /// Creates a new [`FuncType`] from the given `params` and `results`.
+    pub fn new<P, R>(params: P, results: R) -> Self
+    where
+        P: IntoIterator<Item = ValType>,
+        R: IntoIterator<Item = ValType>,
+    {
+        let mut buffer = params.into_iter().collect::<Vec<_>>();
+        let len_params = buffer.len();
+        buffer.extend(results);
+        Self {
+            params_results: buffer.into(),
+            len_params,
+        }
+    }
+
+    /// Creates a new [`FuncType`] fom its raw parts.
+    ///
+    /// # Panics
+    ///
+    /// If `len_params` is greater than the length of `params_results` combined.
+    pub(crate) fn from_raw_parts(params_results: Box<[ValType]>, len_params: usize) -> Self {
+        assert!(len_params <= params_results.len());
+        Self {
+            params_results,
+            len_params,
+        }
+    }
+
+    /// Returns a shared slice to the parameter types of the [`FuncType`].
+    #[inline]
+    pub fn params(&self) -> &[ValType] {
+        // Safety: `len_params` is asserted to be less than or equal to
+        //         `params_results.len()` by construction.
+        unsafe { self.params_results.get_unchecked(..self.len_params) }
```

Comment:
Initially I did not use `unsafe` for `params` and `results` methods but given the new API and its invariants the usage of `unsafe` methods yields slight performance boost for the optimized profile.
Still I wonder if we want to introduce `unsafe` here.

Reply:
Ideally we wouldn't have any `unsafe` code in `wasmparser`, so could you benchmark and see the effect with/without this?

Reply:
Yeah I benchmarked this while working on the PR and the `unsafe` is not critical but just nice to have since it improves some `parse` benchmarks by roughly 2.5%. There is no significant difference between `main` branch and this PR when using safe versions of these methods. So the gain is not big and we can definitely remove the usage of `unsafe` here.

Reply:
**edit:** I just benchmarked it again (since there have been some changes in between) and the performance is actually affected negatively under the default profile without those `unsafe` operations ... It kind of makes sense since the previous implementation naturally also had no bounds checking when both `params` and `returns` was still split.

Reply:
At least for me locally the difference in benchmarks all looks to be noise, I'm measuring +/- 3% with this change. I don't think that these bounds checks are so common that it would actually have any significant impact on performance.

Reply:
Yeah wins and losses are definitely <3% overall. I more or less copied Criterion's meaning about noise or significance. :sweat_smile: 
In summary: With `unsafe` we win 0-3%, without we seems to be losing 0-3% on `parse` benchmarks.

Reply:
At least for me successive runs of `cargo bench` even with the same source code produces +/- 3% variance so anything under 5 I basically don't trust as being accurate personally.

Reply:
That's very interesting since there is a lot less variance on my machine. But yeah, I definitely agree with you in that case.

Reply:
Ok I'm gonna go ahead and merge this since I think the memory wins are still good, but we can keep an eye on this over time. It's not 100% crucial that this crate has 0 unsafe blocks but given the context of `wasmparser` the bar is set pretty high for adding unsafe blocks

Reply:
sounds good to me!
maybe we can add some more different benchmarks in future that can assert certain performance properties better than the currenr set.